### PR TITLE
Product page

### PR DESCRIPTION
**Pull request by samdutton**
Product pages are now working for URLs such as [foo.com/Google+Bot](foo.com/Google+Bot). Any URL that includes a plus sign leads to a product page. A 404 response is returned for non-existent products (e.g. foo.com/Google+Xot].

Still to do: 
• Add to cart
• Product image carousel
• Sizing, number in stock

### CODE REVIEW

**Comment by devnook** [comment]
Sam, can you please rebase to current master?

**Comment by samdutton** [comment]
@devnook I've sorted out the merge problems.

I think we should stick with `node import_products.js` and `node import_home.js` in order to run and merge this branch, then work out what we want we want to do about import/export and data formats.

File: cache/cache [hunk 1]
```diff
@@ -0,0 +1 @@
...=\"nest\" alt=\"Nest logo\" srcset=\"\"></img>\n  </a>\n</section>\n\n</main>\n\n</body>\n\n<footer>\n\n  <p>\n  <a href=\"https://google.com/about\" title=\"About Google\">About Google</a> •\n  <a href=\"http://www.google.com/intl/en/policies/privacy/\" title=\"Privacy Policy\">Privacy</a> •\n  <a href=\"https://shop.googlemerchandisestore.com/store-policies/terms-of-use/\" title=\"Terms of Use\">Terms of Use</a>\n  </p>\n\n  <p><strong>Email us:</strong> <a href=\"mailto:fu@bar.com?subject=Enquiry from Pie Shop demo\" title=\"Email link\">fu@bar.com</a></p>\n\n  <p><strong>Call us:</strong> <a href=\"tel:1-234-567-890\" title=\"Telephone link\">1-234-567-890</a></p>\n\n  <p>Monday to Friday: 8:00am to 8:00pm EST</p>\n\n  <p>\n  <a href=\"https://shop.googlemerchandisestore.com/store-policies/frequently-asked-questions/\" title=\"Shop frequently asked questions\">FAQ</a> • \n  <a href=\"https://shop.googlemerchandisestore.com/store-policies/return-policy/\" title=\"Returns and exchanges\">Returns and Exchanges</a> •\n  <a href=\"https://shop.googlemerchandisestore.com/store-policies/shipping-information/\" title=\"Shipping information\">Shipping</a>\n  </p>\n\n</footer>\n</html>"}
```
[the remark this answers is not in the document]
Comment (samdutton) [reply cache/cache]: Done.

File: package.json [hunk 2]
```diff
@@ -51,10 +51,10 @@
     "test:static": "./node_modules/.bin/karma start",
     "tools:import_products": "./node_modules/.bin/babel-node tools/import_products.js",
     "watch": "npm-run-all --parallel watch:static watch:node",
-    "watch:node": "./node_modules/.bin/watch 'npm run build:node' src/server",
+    "watch:node": "./node_modules/.bin/watch 'npm run build:node' src/server src/shared/js src/data",
```
[the remark this answers is not in the document]
Comment (samdutton) [reply package.json]: Removed.

File: src/server/get-data.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+const fei = require('firestore-export-import');
+fei.initializeApp(process.env.FB_KEYS,
+  'https://firebase.corp.google.com/u/0/project/pie-shop-app/');
+
+let homeData;
+let productData;
+
+export function getHomeData() {
+  if (homeData) {
+    // console.log('>>>>>> had homeData', homeData);
```
[the remark this answers is not in the document]
Comment (samdutton) [reply src/server/get-data.js]: Done.

File: src/server/get-data.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    return homeData;
+  } else {
+    fei.backup('home').then((data) => {
+      // console.log('>>>>>> got homeData', data);
```
[the remark this answers is not in the document]
Comment (samdutton) [reply src/server/get-data.js]: Done.

[the remark this answers is not in the document]
Comment (samdutton) [reply src/server/get-data.js]: BTW — get-data.js should probably live somewhere else.

File: src/server/get-data.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      return homeData;
+    });
+  }
+}
+
+export function getProducts(category) {
+  if (productData) {
+    return filterCategory(productData, category);
+  } else {
+    fei.backup('products').then((data) => {
```
[the remark this answers is not in the document]
Comment (samdutton) [reply src/server/get-data.js]: The aim here is to get the data then cache it, as soon as possible when the user arrives at the site (wherever that may be).

I found this to be much faster than waiting until opening a category page to load data.

We need to work out a strategy for updating data while the user is on the site, but that's another story.

[the remark this answers is not in the document]
Comment (devnook) [reply src/server/get-data.js]: I think it does, but I agree we can cook up something ourselves - this way we are not bound to vendor code for this.

**Pull request merged**